hazelcast/src/main/java/com/hazelcast/internal/usercodedeployment/impl/ClassLocator.java

```diff
@@ -127,65 +134,97 @@ public ClassSource run() {
                     }
-                    return classSource.getClazz();
+                } else {
+                    classSource = new ClassSource(parent, this);
                 }
-                byte[] classDef = fetchBytecodeFromRemote(name);
-                if (classDef == null) {
+                ClassData classData = fetchBytecodeFromRemote(name);
+                if (classData == null) {
                     throw new ClassNotFoundException("Failed to load class " + name + " from other members.");
                 }
-                return defineAndCacheClass(name, classDef);
+
+                Map<String, byte[]> classDefinitions = classData.getClassDefinitions();
+                classSource.define(name, classData.getRequestedClassDefinition());
+                for (Map.Entry<String, byte[]> entry : classDefinitions.entrySet()) {
+                    classSource.define(entry.getKey(), entry.getValue());
+                }
+                cacheClass(classSource, outerClassName);
+                return classSource.getClazz(name);
             }
         } finally {
             IOUtil.closeResource(classMutex);
         }
     }
 
     private Class<?> tryToGetClassFromLocalCache(String name) {
-        ClassSource classSource = classSourceMap.get(name);
+        String parentClassName = extractOuterClassName(name);
+        ClassSource classSource = classSourceMap.get(parentClassName);
         if (classSource != null) {
-            if (logger.isFineEnabled()) {
-                logger.finest("Class " + name + " is already in a local cache. ");
+            Class clazz = classSource.getClazz(name);
+            if (clazz != null) {
+                if (logger.isFineEnabled()) {
+                    logger.finest("Class " + name + " is already in a local cache. ");
+                }
+                return clazz;
             }
-            return classSource.getClazz();
         }
 
-        classSource = clientClassSourceMap.get(name);
+        classSource = clientClassSourceMap.get(parentClassName);
         if (classSource != null) {
-            if (logger.isFineEnabled()) {
-                logger.finest("Class " + name + " is already in a local cache. ");
+            Class clazz = classSource.getClazz(name);
+            if (clazz != null) {
+                if (logger.isFineEnabled()) {
+                    logger.finest("Class " + name + " is already in a local cache. ");
+                }
+                return clazz;
             }
-            return classSource.getClazz();
         }
 
-        classSource = ThreadLocalClassCache.getFromCache(name);
+        classSource = ThreadLocalClassCache.getFromCache(parentClassName);
         if (classSource != null) {
-            return classSource.getClazz();
+            return classSource.getClazz(name);
         }
         return null;
     }
 
+    static String extractOuterClassName(String className) {
+        if (!isInnerClass(className)) {
```

Comment:
Is this check needed? `matcher.matches()` already does that check, If you write as;
```java
       if (matcher.matches()) {
             return matcher.group(1);
        }
        return className;
```